src/main/java/org/cryptomator/cloudaccess/MetadataRequestAggregator.java

```diff
@@ -32,15 +33,17 @@
 	private final CloudProvider delegate;
 
 	public MetadataRequestAggregator(CloudProvider delegate) {
-		this(delegate, Duration.ofSeconds( //
-				Integer.getInteger("org.cryptomator.cloudaccess.metadatacachingprovider.timeoutSeconds", DEFAULT_CACHE_TIMEOUT_SECONDS) //
-		));
+		this(
+				delegate, //
+				Duration.ofSeconds(Integer.getInteger("org.cryptomator.cloudaccess.metadatacachingprovider.timeoutSeconds", DEFAULT_CACHE_TIMEOUT_SECONDS)), //
+				Duration.ofSeconds(Integer.getInteger("org.cryptomator.cloudaccess.metadatacachingprovider.aggregateRequestsMaxAgeSeconds", DEFAULT_AGGREGATE_REQUESTS_MAX_AGE)) //
```

Comment:
In my opinion, this is a misleading configuration name. If I'm not mistaken, the purpose of this class is _not only_ the **new** feature to avoid duplicate requests while a request is still running, but still to cache metadata responses (and now also list responses).

Even if we received a response, it makes sense to keep the completed future in our cache. After all the purpose of this caching layer is to not only avoid request while waiting for a response, but also a few seconds after.

So my suggestion is:
1. Either use `org.cryptomator.cloudaccess.metadatacachingprovider.timeoutSeconds` for all three caches.
2. Or use three separate config properties, i.e. `metadataTimeoutSeconds`, `itemListTimeoutSeconds`, `quotaTimeoutSeconds`

Reply:
> If I'm not mistaken, the purpose of this class is not only the new feature to avoid duplicate requests while a request is still running, but still to cache metadata responses (and now also list responses).
> 
> Even if we received a response, it makes sense to keep the completed future in our cache. After all the purpose of this caching layer is to not only avoid request while waiting for a response, but also a few seconds after.

We now use the `MetadataRequestAggregator` regarding `itemMetadata` and `list` only for aggregating responses so that a list request which is already running not executed a second and a "thousand" time within a second as these methods are often called very quickly one after the other on the same path but as soon as the response is available, the caching is handled by our `NodeCache` in the `CachedPropfindEntryProvider`. The `NodeCache` knows due to `ETag` the best if a cached entry is still valid that is why the response in the `MetadataRequestAggregator` should be in my opinion invalidated as soon as possible so that we have only one cache and one aggregator running and not two caches.

Reply:
Oh ok I see. While this is capable of cashing, it is not intended to act as a cache (at least not if the backend supports etags).

So the question is if we want it to show this capability or not. If not, we should move the quota thing to a different place, shouldn't we?

Reply:
We can go even further and remove the expiration from the `cachedItemMetadataRequests` and `cachedItemListRequests` completely but invalidate it as soon as the response is available, IMO this is the correctest version, e.g. `itemMetadata` would look like this:

```java
return cachedItemMetadataRequests
    .get(node, () -> delegate.itemMetadata(node)
    .whenComplete((metadata, throwable) -> invalidateAggregators(node)));
```
Then we have no overlapping if e.g. the response isn't fast enough and the functionality of it is clearer.

---

Regarding the quota cache I totally agree with you, quota is somehow wrong here.

---

I think we should
1. As mentioned above, remove the expiration from `cachedItemMetadataRequests` and `cachedItemListRequests` and invalidate immediately after receiving the response
2. Create a new `QuotaCachingProviderDecorator` which caches `quota` requests
3. Re-Implement the previous functionality of the `MetadataCachingProviderDecorator` in case of a cloud doesn't support caching like WebDAV without ETags so that the cloud state is present after a request until the entry gets invalidated e.g. due to the timeout. But use this inside of e.g. the `WebDAVProvider` as we don't know from the outside if we can use caching.

Reply:
Moving from cache to a simple map + manual removal was my first thought, too.

My second thought, though, was: Would it hurt to keep this as a second caching layer? Thinking of servers without etag support here.

Reply:
`expireAfterWrite(0)` looks good:

> When duration is zero, this method hands off to maximumSize(0), ignoring any otherwise-specificed maximum size or weight. This can be useful in testing, or to disable caching temporarily without a code change. 

`maximumSize(0)` in return leads to

> ...elements will be evicted immediately after being loaded into cache. This can be useful in testing, or to disable caching temporarily without a code change. 

I'll implement it this way :+1:

Reply:
Needs still some further unit tests.

Reply:
FYI, you might benefit from using [Caffeine](https://github.com/ben-manes/caffeine) here instead of Guava due to native support for futures via [AsyncCache](https://github.com/ben-manes/caffeine/wiki/Population#asynchronous-manual). An in-flight future will not be eligible for eviction by a maximum size or expiration policy, and will be automatically removed if the future fails or resolves to a `null` value. That might simplify your code, but I couldn't tell if you wanted to cache exceptions or not.

The comments about preventing duplicate requests with immediate eviction would in your Guava code mean only for the duration of creating the future (usually fast), whereas in Caffeine it would be until the future completed. I suspect the latter is what you'd want.

Reply:
@ben-manes That does indeed look interesting (for various applications). We'll have a look at it, thanks for the hint!

Reply:
>  you might benefit from using Caffeine here instead of Guava due to native support for futures via AsyncCache. 

Even Google states ["Prefer Caffeine over Guava's caching API"](https://guava.dev/releases/snapshot/api/docs/com/google/common/cache/CacheBuilder.html), thanks for your input @ben-manes, it looks really great and we can definitely use Caffeine in other places. Here I'm still evaluating, but I don't think so, because under certain circumstances (the cloud or a server doesn't support smarter caching) we actually want to prevent not only duplicates on this level but really cache, which includes exceptions.

> The comments about preventing duplicate requests with immediate eviction would in your Guava code mean only for the duration of creating the future (usually fast), whereas in Caffeine it would be until the future completed. I suspect the latter is what you'd want.

As we (currently) don't set an expiration duration but invalidate manually, this doesn't apply AFAIK (and tested it):

> These features are all optional; caches can be created using all or none of them. By default cache instances created by CacheBuilder will not perform any type of eviction. (from https://guava.dev/releases/18.0/api/docs/com/google/common/cache/CacheBuilder.html)